policies/documentation-policy.md

```diff
@@ -0,0 +1,123 @@
+# OpenSSL documentation policy
+
+This document describes the code documentation and commenting
+requirements for the OpenSSL project.
+
+The project's documentation is all about making the libraries and tools
+more accessible to our users and making the code more maintainable.
+This policy applies to new submissions, existing code is below par and will
+be gradually improved.
+
+## Chapter 1: Command line commands and arguments
+
+All new or modified arguments to the commands must be documented in the
+`doc/man1` directory.  This documentation is in _POD_ format.
+
+All new commands must be documented in the `doc/man1` directory.  This
+documentation is in _POD_ format.
+
+## Chapter 2: Public symbols in the libraries
+
+All new public symbols must be documented in a _POD_ manual page in the
+`doc/man3` directory.  This includes both macros and functions.
+
+The allowed exceptions are:
+
+- guard macros preventing a header file being included twice
+- new OPENSSL_NO_DEPRECATED_ macros
+- new symbols generated automatically via `make update` (errors, objects, etc)
+
+## Chapter 3: Overviews, conventions, et al
+
+Where additional user facing information is required, it should be
+included in the `doc/man7` section.  This includes, but is not limited to:
+
+- descriptions of data structures and their usage
+- provider calls and call backs
+- algorithm descriptions and parameters
+- descriptions of architectural decisions
+
+## Chapter 4: Internal functions, structures, globals and macros
+
+Internal functions and macros are those defined in the `include/crypto`
+and `include/internal` directories.
+
+These should all be documented at the point of implementation.  A comment
+describing their purpose and the input and output arguments and return value
+for functions suffices.
+
+For _trivial_ items, where their operation is obvious from their
+implementation, the documentation requirement is not required.  The following
+are generally representative of trivial items, however it is quite
+possible for any of these to be non-trivial in specific instances and
+therefore require documentation:
+
+- OSSL_DISPATCH tables
+- upref functions
+- free functions
+- simple getter/setter functions
+- wrappers for other functions (a function that calls a more recent `_ex`
+  variant or a group of functions that call a common internal routine)
+
+For structures, each of the fields should be commented stating its
+purpose.  Again, a _trivial_ exception applies where the purpose is
+obvious.  Some representative examples:
+
+- `OSSL_LIB_CTX *ctx;` where there is only one library context referenced in
+  the structure.
+- `struct *next;` in a linked list implementation.
+- `CRYPTO_REF_COUNT refcnt;`
+- `CRYPTO_RWLOCK *lock;` where there is only one lock in the structure and
```

Comment:
I actually think a lock is virtually never trivial. Even in the case where there is only one and it locks the whole structure, I think it is necessary to explicitly state that it locks the whole structure.

Reply:
Removed.